feat: New page for UUIDs (#5537)
JetBrains/kotlin-web-site: docs/topics/uuids.md

[//]: # (title: UUIDs)
[//]: # (description: Learn how to use UUIDs in Kotlin, including creating, parsing, formatting, serializing, and working with UUID values across multiplatform and JVM code.)

The [`Uuid`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/)
class represents universally unique identifiers (UUIDs),

> also known as Globally Unique Identifiers (GUIDs).

Reviewer: ```suggestion
also known as globally unique identifiers (GUIDs).
```

Reviewer: since we didn't capitalize it above

Author: Nice catch! But prefer to capitalize the first instance instead.

Reviewer: also works 👍  thank you

A `Uuid` is a 128-bit value used to uniquely identify an entity without
relying on a central system that assigns IDs. This makes UUIDs useful
in distributed applications, databases, client-generated records,
or [Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform/get-started.html) applications.

Use the `Uuid` class to work with UUID values. Unlike plain strings, a dedicated UUID
type makes your code more explicit and prevents accidental use of invalid values.

> The Kotlin standard library provides the `Uuid` class in the `kotlin.uuid` package:

Reviewer: this feels a bit unclear WDYT about: To use UUIDs in your project, import the `Uuid` class from the `kotlin.uuid` package:

Author: Yes, sounds much better!

```kotlin
import kotlin.uuid.Uuid
```

> ## Generate UUIDs

Reviewer: ```suggestion

## Generate UUIDs
```

Reviewer: missing line

> To create a new random UUID, use the [`Uuid.random()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/random.html) function.

Reviewer: ```suggestion
To generate a new random UUID, use the [`Uuid.random()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/random.html) function.
```

Reviewer: to make it consistent with the title

> It generates a version 4 UUID from random values.

Reviewer: I think we could make this a bit more concise, right now I feel like it repeats the same idea twice: 

To generate a random version 4 UUID for regular identifiers, such as user or database IDs, use the [`Uuid.random()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/random.html) function:

Use the `Uuid.random()` function for regular identifiers, such as user or database IDs.

```kotlin
import kotlin.uuid.Uuid

> fun main() {

Reviewer: ```suggestion

fun main() {
```

//sampleStart    
    val id = Uuid.random()
    println(id)
//sampleEnd    
}
```

> {kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

Reviewer: ```suggestion
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}
```

Reviewer: afaik this only works from 2.0.0 and up but I would omit it because some of these APIs only work 2.3.0 and up.

Author: Nice catch! Will apply where possible.

> If your application require a specific version of UUIDs,

Reviewer: ```suggestion
If your application requires a specific version of UUIDs,
```

Reviewer: WDYT about making this list a bit less crowded we could perhaps introduce this part with something like.

> If your application requires a specific UUID version, you can generate UUIDs with version-specific functions.

Then we could follow it up with 
Use the `Uuid.generateV4()` function to generate a version 4 UUID:

[example]

This function makes the UUID version explicit, but otherwise works the same as `Uuid.random()`.

To generate version 7 UUID... and so on

you can use the following functions:

* The [`Uuid.generateV4()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v4.html) function creates the same type of UUID as the `Uuid.random()` function
  but explicitly states that the value is a version 4 UUID.

  Use it when version 4 UUIDs are specifically required.

  ```kotlin
  import kotlin.uuid.Uuid

>   fun main() {

Reviewer: same comments here, please add a line after import and change the version - here it's only 2.3 afaik but I suggest omitting it.

  //sampleStart    
      val id = Uuid.generateV4()
      println(id)
  //sampleEnd    
  }
  ```
  {kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

* The [`Uuid.generateV7()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7.html) function creates a version 7 UUID that contains
  a time-based component and random data.

  Use it when you need to sort UUIDs by creation time:

  ```kotlin
  import kotlin.uuid.Uuid
  fun main() {
  //sampleStart    
      val id = Uuid.generateV7()
      println(id)
  //sampleEnd    
  }
  ```
  {kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

>   You can also generate a version 7 UUID for a specific moment in time,

Reviewer: ```suggestion
  You can also generate a version 7 UUID for a specific moment in time
```

>   You can also generate a version 7 UUID for a specific moment in time,

Reviewer: I think we could add an example here as well. 

Also I suggest adding information about `Uuid.generateV7NonMonotonicAt()` not guaranteeing monotonic ordering, so multiple UUIDs created for the same timestamp might not be sequential.

Author: The use case for this function is quite niche tbh. I don't think it's worth documenting it here - people who may need it, would be able to do so using API ref.

>   using the [`Uuid.generateV7NonMonotonicAt()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7-non-monotonic-at.html) function.

Reviewer: ```suggestion
  with the [`Uuid.generateV7NonMonotonicAt()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7-non-monotonic-at.html) function.
```

> > Learn more about the generation algorithm in the [API Reference](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/).

Reviewer: I think this tip is a bit narrow if we are linking to the companion object. 
If we want to provide more information, I suggest something like

 > For the complete list of `Uuid` companion functions, see the [`Uuid.Companion` API reference](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/).

But even that's a bit too narrow because then we don't talk about properties, so I would suggest omitting it, we already linked to the relevant parts above.

Author: fair enough!

>
{style="tip"}

## Parse UUIDs

> Some UUIDs can be first declared as strings.

Reviewer 2: can be represented as strings?

Author: thanks, good catch, updated!

> For example, when you receive a UUID from a URL parameter or database record.

Reviewer: I think we could make this a bit more compact: 

UUID values are often represented as strings, such as in URL parameters or database records.

To convert a `String` value to a `Uuid` value,
use the [`Uuid.parse()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse.htm) function:

```kotlin

> val id = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")

Reviewer: could we make this into a runnable example as well? 🤔 — it feels weird that the runnable example is the one that isn't suppoed to work.

Author: Converted to align, but I'm not sure if it is needed, since the output looks just like the content we have in (). It does not demonstrate the change from string to a Uuid.

```

The `Uuid.parse()` function accepts both the standard hex-and-dash
format and hexadecimal format without dashes.

If the input is invalid, the `Uuid.parse()` function throws an `IllegalArgumentException`:

```kotlin
import kotlin.uuid.Uuid
fun main() { 
//sampleStart    
    val id = Uuid.parse("10")
    println(id)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" validate="false"}

> If your application accepts only one representation, use:

Reviewer: ```suggestion
If your application accepts only one representation, use the format-specific functions:
```

Reviewer: this way we could make the list a bit more concise: 

* [`Uuid.parseHexDash()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash.html) for the hex-and-dash string representation.
* [`Uuid.parseHex()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex.html) for the hexadecimal string representation without dashes.

> * The [`Uuid.parseHexDash()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash.html) function to parse from the hex-and dash string representation.

Reviewer: ```suggestion
* The [`Uuid.parseHexDash()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash.html) function to parse from the hex-and-dash string representation.
```

* The [`Uuid.parseHex()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex.html) function to parse from the hexadecimal string representation without dashes.

For example:

> ```kotlin

Reviewer: we could also make this runnable WDYT?

val standard = Uuid.parseHexDash("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
val compact = Uuid.parseHex("de2bc56cea734f3c8a375a46fdb2d79a")
```

> If you have UUIDs from external input, use the

Reviewer: I think we can refine that and fold the tip into it. WDYT:

If your application handles UUID strings from external sources and must handle invalid input without throwing exceptions, use `Uuid.parseOrNull()`, `Uuid.parseHexDashOrNull()`, or `Uuid.parseHexOrNull()`. These functions return `null` when the input is invalid:

[Example]

[`Uuid.parseOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-or-null.html), [`Uuid.parseHexDashOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash-or-null.html),
or [`Uuid.parseHexOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-or-null.html)
functions to handle invalid values safely:

```kotlin

> fun parseId (input: String): Uuid? {

Reviewer: we could also make this runnable 🏃

    return Uuid.parseOrNull(input)
}
```

> > Use the `Uuid.parseOrNull()`, `Uuid.parseHexDashOrNull()`, and `Uuid.parseHexOrNull()` functions

Reviewer: I think we can explain this in flow text instead of making it a tip see my comment above

> when invalid input is part of your normal application workflow.
>
{style="tip"}

> ## Convert UUIDs

Reviewer: Let's make the title a bit more specific WDYT? if we are only converting to strings something like : 

Convert UUIDs to strings

> You can convert a `Uuid` value to a `String` value using the following functions:

Reviewer: ```suggestion
You can convert a `Uuid` value to a `String` value with the following functions:
```

> * [`.toString()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-string.html) for the standard string representation

Reviewer: afaik these are member functions not extension functions so we don't need the `.˙ in `toString()` here and for the rest in this list

Author: nice catch!

* [`.toHexDashString()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-hex-dash-string.html) for the hex-and-dash format
* [`.toHexString()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-hex-string.html) for the hexadecimal format without dashes

For example:

```kotlin
import kotlin.uuid.Uuid
fun main() {
//sampleStart    
    val id = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    

>     println(id.toString()) // de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a

Reviewer: ```suggestion
    println(id.toString())
    // de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a
```

Reviewer: If the code outputs a result , reflect this in a comment on the next line.

    println(id.toHexDashString()) // de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a
    println(id.toHexString()) // de2bc56cea734f3c8a375a46fdb2d79a 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

## Compare UUIDs

> You can perform equality checks on the `Uuid` values using the `==` operator.

Reviewer: 🤔 this one parses a bit mechanical WDYT something like: You can check whether `Uuid` values are equal using the `==` operator.

> Kotlin evaluates values according to the UUID value, not to the textual representation.

Reviewer: here I think we could make the connection clearer by saying something along the lins of: 
Kotlin compares values according to the UUID value, not the textual representation. 

Kotlin evaluates values according to the UUID value feels a bit more distant compared to compares

> For example, two values in the different forms are equal if they represent the same 128-bit value:

Reviewer: ```suggestion
For example, two values in different forms are equal if they represent the same 128-bit value:
```

```kotlin
import kotlin.uuid.Uuid
fun main() {
//sampleStart    
    val first = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    val second = Uuid.parse("de2bc56cea734f3c8a375a46fdb2d79a")

    println(first == second) // true 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> This functionality makes the `Uuid` comparison safer than comparing UUIDs as strings.

Reviewer: ```suggestion
This makes the `Uuid` comparison safer than comparing UUIDs as strings.
```

Reviewer: do we need this part? 🤔

Reviewer: I feel like the word "safer" is a bit misleading here. It could probably give you a false negative, but wouldn't crash anything afaik. perhaps reliable? 🤔 WDYT?

> String comparison fails if you write the same value in different formats.

Reviewer: saying it fails is a bit strange here 🤔  The comparison doesn't fail - it returns `false`. 
so wdyt: String comparison treats the same value in different formats as different values.

> `Uuid` comparison checks the actual identifier value.

Reviewer: Hmm 🤔 reading these sentences together, WDYT about merging these 3 sentences into one:

This makes `Uuid` comparison more reliable than string comparison, which treats different string formats as different values.

> You can also sort UUIDs using the `Comparable<Uuid>` interface. In that case,

Reviewer: I might be missing a connection here 🤔 Could you please explain how `Comparable` connects to sorting in relation to UUIDs? 
Do we mean that because `Uuid` implements `Comparable<Uuid>`, the user can sort UUIDs with standard collection functions such as `sorted()`.

Author: Exactly, the connection is:

Uuid implements Comparable<Uuid>, which means UUID values have a defined natural order. Because of that, collection functions that rely on natural ordering, such as sorted(), can sort a collection of UUIDs directly.

Rephrased to make clearer.

Kotlin compares values lexicographically (from the most to the least significant bit):

```kotlin
import kotlin.uuid.Uuid
fun main() {
//sampleStart    
    val first = Uuid.generateV7()
    val second = Uuid.generateV7()

    val sorted = listOf(first, second).sorted()
    println(sorted) 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> ## Work with bytes

Reviewer: ```suggestion
## Work with UUID binary representations
```

Reviewer: WDYT would that work? 🤔

Kotlin provides functionality to work with the 128-bit value of a UUID.
Use it when you need binary storage, custom serialization, or interoperability with other systems.

To convert a UUID to and from 16-byte representation, use the [`.toByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-byte-array.html)
and [`Uuid.fromByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/from-byte-array.html) functions:

```kotlin
import kotlin.uuid.Uuid
fun main() {
//sampleStart 
    val id = Uuid.random()

    val bytes = id.toByteArray()
    val original = Uuid.fromByteArray(bytes)
  
    println(id)
    
    println(bytes)
    println(original)

>     println(id == original) // true

Reviewer: ```suggestion
    println(id == original)
    // true
```

//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> You can also create a UUID from two 64-bit values. Since Kotlin does not provide a built-in 128-bit integer type,

Reviewer: ```suggestion
You can also create a UUID from two 64-bit values. Since Kotlin doesn't provide a built-in 128-bit integer type,
```

> You can also create a UUID from two 64-bit values. Since Kotlin does not provide a built-in 128-bit integer type,

Reviewer: I found this transition really difficult to follow. 
Do you think we could start the section to introduce that these are basically two ways to do this: 

Some APIs and storage formats represent UUIDs as bytes or as two 64-bit values. representing bytes this way --- add use cases
Then we can follow with .toByteArray() example. 

Then transition to this one. with the explanation that Kotlin doesn't provide a built-in 128-bit integer type

Author: Restructured. WDYT?

the API represents a UUID as two `Long` values using:
* The `mostSignificantBits` parameter for the first 64 bits of a UUID.
* The `leastSignificantBits` parameter for the last 64 bits of a UUID.

> If another system gives you a UUID in the two-part format,

Reviewer: does it have to be another system ? 🤔 Could something like this stand on its own?

To create a `Uuid` value from two `Long` values, use the [`Uuid.fromLongs()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/from-longs.html) function:

use the [`Uuid.fromLongs()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/from-longs.html) function to create a single `Uuid` value:

```kotlin
import kotlin.uuid.Uuid
fun main() {
//sampleStart 
    val id = Uuid.fromLongs(
        mostSignificantBits = -4653685776373167443,
        leastSignificantBits = -6288180676521310383.toLong()
    )
    println(id) // bf6ac971-52fd-4aad-a8bb-e4fdac78c751
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

To extract two parts from an existing `Uuid` value, use the [`Uuid.toLongs()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-longs.html) function:

```kotlin
import kotlin.uuid.Uuid
fun main() {
//sampleStart 
    val id = Uuid.random().toLongs { mostSignificantBits, leastSignificantBits ->
    println(mostSignificantBits)
    println(leastSignificantBits)
    }
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

## Serialize UUIDs

> Kotlin supports serialization for the `Uuid` values.

Reviewer: ```suggestion
Kotlin supports serialization for `Uuid` values.
```

Use it to store or transfer a UUID value outside Kotlin code, for example, in JSON APIs or configuration files.

> To serialize a `Uuid` value, represent it as a `String` value unless otherwise required by your system.

Reviewer: ```suggestion
To serialize a `Uuid` value, represent it as a string unless your application requires another format.
```

> The [`kotlinx.serialization`](https://kotlinlang.org/docs/serialization.html) functionality uses the standard hex-and-dash format:

Reviewer: ```suggestion
The [`kotlinx.serialization`](https://kotlinlang.org/docs/serialization.html) library uses the hex-and-dash format:
```

Reviewer: I wouldn't call this a functionality, it's a library/libraries (or a plugin depending on the context) 

regarding standard: is there a reason to call it standard hex-and-dash? 🤔

```kotlin
//sampleStart 
import kotlin.uuid.Uuid
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class User(
    val id: Uuid,
    val name: String
)

fun main() {
    val user = User(
        id = Uuid.parse("de2bc56cea734f3c8a375a46fdb2d79a"),
        name = "Kotlin"
    )

    println(Json.encodeToString(user))
    // {"id":"de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a","name":"Kotlin"}
}
//sampleEnd
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> ## Use UUIDs on the JVM

Reviewer: Hmm 🤔 I feel like the section focuses more on the interaction between java.util.UUID and Kotlin UUIDs than on the JVM in my opinion, WDYT about a title like ## Use UUIDs with Java APIs or similar

> When you work on the JVM, you can interoperate your Kotlin code with Java libraries

Reviewer: Do you think we could make this a bit more concise? 

Someting like: 

Java represents UUIDs with the `java.util.UUID` class. On the JVM, Java APIs may accept or return this type.

Although `java.util.UUID` and `kotlin.uuid.Uuid` both represent UUIDs, they are separate classes/types. 
(we could even add the note from below here)

To convert Java UUID to Kotlin, use X..

that use the `java.util.UUID` class.

The `java.util.UUID` and `kotlin.unit.UUID` classes represent the same kind of values,
but they are two distinct types. You need to convert values explicitly to pass UUIDs between Kotlin and Java:

> * Convert a Java UUID to Kotlin using the [`.toKotlinUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-kotlin-uuid.html) function:

Reviewer: ```suggestion
* Convert a Java UUID to Kotlin with the [`.toKotlinUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-kotlin-uuid.html) extension function:
```

class represents Universally Unique Identifiers (UUIDs),
also known as Globally Unique Identifiers (GUIDs).

A `Uuid` is a 128-bit value used to uniquely identify an entity without
relying on a central system that assigns IDs. This makes UUIDs useful
in distributed applications, databases, client-generated records,
or [Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform/get-started.html) applications.

Use the `Uuid` class to work with UUID values. Unlike plain strings, a dedicated UUID
type makes your code more explicit and prevents accidental use of invalid values.

To use UUIDs in your project, import the `Uuid` class from the `kotlin.uuid` package:

```kotlin
import kotlin.uuid.Uuid
```

## Generate UUIDs

To generate a random version 4 UUID for regular identifiers, such as user or database IDs, 
use the [`Uuid.random()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/random.html) function:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart    
    val id = Uuid.random()
    println(id)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

> You can also generate a specific version of UUIDs with the following functions:

Reviewer: ```suggestion
You can also generate specific versions of UUIDs with the following functions:
```

Reviewer: or You can also generate UUIDs of specific versions

* The [`Uuid.generateV4()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v4.html) function generates the same type of UUID as the `Uuid.random()` function 
  but explicitly states that the value is a version 4 UUID.

> *  The [`Uuid.generateV7()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7.html) function generates a version 7 UUID with a timestamp

Reviewer: ```suggestion
* The [`Uuid.generateV7()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7.html) function generates a version 7 UUID with a timestamp
```

>    that you can use for UUIDs sorting.

Reviewer: ```suggestion
   that you can use for sorting UUIDs.
```

Reviewer: or for UUID sorting

* The [`Uuid.generateV7NonMonotonicAt()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7-non-monotonic-at.html) function generates a version 7 UUID for a specific moment in time.

Here's an example that generates version-specific UUIDs:

```kotlin
import kotlin.time.Instant
import kotlin.uuid.Uuid
  
fun main() { 
//sampleStart    

>     val idVersion4 = Uuid.generateV4()

Reviewer: Could you please add code comments here before each relevant line like val idVersion4 = ...? 

// Generates a version 4 UUID
// Generates a version 7 UUID
// Generates a version 7 UUID for the specified timestamp

    println(idVersion4)
    
    val idVersion7 = Uuid.generateV7()
    println(idVersion7)

>    val timestamp = Instant.fromEpochMilliseconds(1757440583000L)

Reviewer: fix indendation in this lines (only 3 spaces instead of 4)

//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

If your application accepts only one representation, use
the format-specific functions:

* [`Uuid.parseHexDash()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash.html) for the hex-and-dash string representation.
* [`Uuid.parseHex()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex.html) for the hexadecimal string representation without dashes.

For example:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart  
    val standard = Uuid.parseHexDash("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    val compact = Uuid.parseHex("de2bc56cea734f3c8a375a46fdb2d79a")
    
    println(standard)
    println(compact)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

If you have UUIDs from external sources and must handle invalid input safely,
use [`Uuid.parseOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-or-null.html), [`Uuid.parseHexDashOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash-or-null.html),

> or [`Uuid.parseHexOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-or-null.html). 

Reviewer: ```suggestion
or [`Uuid.parseHexOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-or-null.html).
```

    println(first == second) 
    // true 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

This makes the `Uuid` comparison more reliable than comparing strings,
which treats the same value in different formats as different values.
`Uuid` comparison checks the actual identifier value.

`Uuid` implements the `Comparable<Uuid>` interface, so UUID values can be sorted with standard collection
functions such as `sorted()`. In that case,
Kotlin compares values lexicographically (from the most to the least significant bit):

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart    
    val first = Uuid.generateV7()
    val second = Uuid.generateV7()

    val sorted = listOf(first, second).sorted()
    println(sorted) 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

## Work with binary representations

Some APIs, storage formats, and binary protocols do not represent UUIDs as strings. 
Instead, they store the 128-bit UUID value as either:

* A 16-byte array 
* Two 64-bit values

Use these representations when you need to exchange UUIDs with systems that expect binary UUID data.

> To convert a UUID to and from a 16-byte representation, use the [`.toByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-byte-array.html)

Reviewer: (stlye suggestion) 
To convert between a UUID and a 16-byte representation — this way we can avoid to and from

    // {"id":"de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a","name":"Kotlin"}
}
//sampleEnd
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

## Use UUIDs with Java APIs

Java uses the `java.util.UUID` class to represent UUIDs. On the JVM, Java APIs may
accept or return this type. Although `java.util.UUID` and `kotlin.uuid.Uuid` both represent UUIDs,
they are two distinct types. 

Convert values explicitly to pass UUIDs between Kotlin and Java:

* Convert a Java UUID to Kotlin with the [`.toKotlinUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-kotlin-uuid.html) extension function:

  ```kotlin
  import kotlin.uuid.toKotlinUuid
  
  val kotlinId: Uuid = javaId.toKotlinUuid()
  ```

* Convert a Kotlin UUID to Java with the [`.toJavaUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-java-uuid.html) extension function:

  ```kotlin
  import kotlin.uuid.toJavaUuid
  
  val javaId: java.util.UUID = kotlinId.toJavaUuid()
  ```

These functions allow you to represent your UUID values using `Uuid` at JVM interoperability boundaries.

> The `java.util.UUID` and `kotlin.uuid.Uuid` classes are comparable, but the ordering may differ.

> > Make sure to check your code that depends on UUID ordering before migrating from Java API to Kotlin API.

Reviewer: 🤔 
If we mean specifically between these two : `java.util.UUID` and `kotlin.uuid.Uuid` 
let's add the artice:   Make sure to check your code that depends on UUID ordering before migrating from the Java API to the Kotlin API.

  that you can use for UUIDs sorting:

  ```kotlin
  import kotlin.uuid.Uuid
  
  fun main() {
  //sampleStart    
      val id = Uuid.generateV7()
      println(id)
  //sampleEnd    
  }
  ```
  {kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

  You can also generate a version 7 UUID for a specific moment in time
  with the [`Uuid.generateV7NonMonotonicAt()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7-non-monotonic-at.html) function.

## Parse UUIDs

UUID values are often represented as strings,
such as in URL parameters or database records.

To convert a `String` value to a `Uuid` value,
use the [`Uuid.parse()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse.html) function:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart
    val id = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    println(id)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

The `Uuid.parse()` function accepts both the standard hex-and-dash

> format and hexadecimal format without dashes.

Reviewer: ```suggestion
format and the hexadecimal format without dashes.
```

    println(first == second) 
    // true 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

This makes the `Uuid` comparison more reliable than comparing strings,
which treats the same value in different formats as different values.
`Uuid` comparison checks the actual identifier value.

`Uuid` implements the `Comparable<Uuid>` interface, so UUID values can be sorted with standard collection
functions such as `sorted()`. In that case,
Kotlin compares values lexicographically (from the most to the least significant bit):

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart    
    val first = Uuid.generateV7()
    val second = Uuid.generateV7()

    val sorted = listOf(first, second).sorted()
    println(sorted) 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

## Work with binary representations

Some APIs, storage formats, and binary protocols do not represent UUIDs as strings. 
Instead, they store the 128-bit UUID value as either:

* A 16-byte array 
* Two 64-bit values

Use these representations when you need to exchange UUIDs with systems that expect binary UUID data.

> To convert a UUID to and from 16-byte representation, use the [`.toByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-byte-array.html)

Reviewer: ```suggestion
To convert a UUID to and from a 16-byte representation, use the [`.toByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-byte-array.html)
```

and [`Uuid.fromByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/from-byte-array.html) functions:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart 
    val id = Uuid.random()

    val bytes = id.toByteArray()
    val original = Uuid.fromByteArray(bytes)
  
    println(id)
    
    println(bytes)
    println(original)

    println(id == original) 
    // true
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

You can also represent the same 128-bit UUID value as two `Long` values. 

> This is useful because Kotlin does not provide a built-in 128-bit integer type. 

Reviewer: ```suggestion
This is useful because Kotlin doesn't provide a built-in 128-bit integer type. 
```

The two `Long` values store the UUID in two parts:

* The `mostSignificantBits` parameter for the first 64 bits of a UUID.
* The `leastSignificantBits` parameter for the last 64 bits of a UUID.

To create a `Uuid` value from two `Long` values,
use the [`Uuid.fromLongs()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/from-longs.html) function:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart 
    val id = Uuid.fromLongs(
        mostSignificantBits = -4653685776373167443,
        leastSignificantBits = -6288180676521310383.toLong()
    )
    println(id) 
    // bf6ac971-52fd-4aad-a8bb-e4fdac78c751
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

> To extract two parts from an existing `Uuid` value, use the [`Uuid.toLongs()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-longs.html) function:

Reviewer: ```suggestion
To extract the two parts from an existing `Uuid` value, use the [`Uuid.toLongs()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-longs.html) function:
```

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart 

>     val id = Uuid.random().toLongs { mostSignificantBits, leastSignificantBits ->

Reviewer: I think we should use something like this: 

```

import kotlin.uuid.Uuid

fun main() {
//sampleStart 
    val id = Uuid.random()

    id.toLongs { mostSignificantBits, leastSignificantBits ->
        println(mostSignificantBits)
        println(leastSignificantBits)
    }
//sampleEnd  
}
```

The issue with the current snippet is that it assigns the result of `toLongs()` to `id`, but the lambda returns `Unit`, so it's not a Uuid.

    println(mostSignificantBits)
    println(leastSignificantBits)
    }
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

## Serialize UUIDs

Kotlin supports serialization for `Uuid` values.
Use it to store or transfer a UUID value outside Kotlin code, for example, in JSON APIs or configuration files.

To serialize a `Uuid` value, represent it as a string unless your application requires another format.
The [`kotlinx.serialization`](https://kotlinlang.org/docs/serialization.html) library uses the hex-and-dash format:

```kotlin
//sampleStart 
import kotlin.uuid.Uuid
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class User(
    val id: Uuid,
    val name: String
)

fun main() {
    val user = User(
        id = Uuid.parse("de2bc56cea734f3c8a375a46fdb2d79a"),
        name = "Kotlin"
    )

    println(Json.encodeToString(user))
    // {"id":"de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a","name":"Kotlin"}
}
//sampleEnd
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

## Use UUIDs with Java APIs

> To represent UUIDs, Java uses `java.util.UUID` class. On the JVM, Java APIs may

Reviewer: ```suggestion
Java uses the `java.util.UUID` class to represent UUIDs. On the JVM, Java APIs may
```

> accept or return this type. Although `java.util.UUID` and `kotlin.unit.Uuid` both represent UUIDs,

Reviewer: ```suggestion
accept or return this type. Although `java.util.UUID` and `kotlin.uuid.Uuid` both represent UUIDs,
```

they are two distinct types. 

Convert values explicitly to pass UUIDs between Kotlin and Java:

* Convert a Java UUID to Kotlin with the [`.toKotlinUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-kotlin-uuid.html) extension function:

  ```kotlin
  import kotlin.uuid.toKotlinUuid
  
  val kotlinId: Uuid = javaId.toKotlinUuid()
  ```

* Convert a Kotlin UUID to Java with the [`.toJavaUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-java-uuid.html) extension function:

  ```kotlin
  import kotlin.uuid.toJavaUuid
  
  val javaId: java.util.UUID = kotlinId.toJavaUuid()
  ```

These functions allow you to represent your UUID values using `Uuid` at JVM interoperability boundaries.

> > The `java.util.UUID` and `kotlin.unit.Uuid` classes are comparable, but the ordering may differ.

Reviewer: ```suggestion
> The `java.util.UUID` and `kotlin.uuid.Uuid` classes are comparable, but the ordering may differ.
```

> Make sure to check your code that depends on UUID ordering before migrating from Java API to Kotlin API.
>
{style="note"}

Kotlin also provides support for working with Java buffers.

> Use JVM-specific functions to work with UUIDs from a `ByteBuffer`:

Reviewer: ```suggestion
Use JVM-specific functions to work with UUIDs in a `ByteBuffer`:
```

Reviewer: nit: from works for get but for put it parses slightly wrong

class represents Universally Unique Identifiers (UUIDs),
also known as Globally Unique Identifiers (GUIDs).

A `Uuid` is a 128-bit value used to uniquely identify an entity without
relying on a central system that assigns IDs. This makes UUIDs useful
in distributed applications, databases, client-generated records,
or [Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform/get-started.html) applications.

Use the `Uuid` class to work with UUID values. Unlike plain strings, a dedicated UUID
type makes your code more explicit and prevents accidental use of invalid values.

To use UUIDs in your project, import the `Uuid` class from the `kotlin.uuid` package:

```kotlin
import kotlin.uuid.Uuid
```

## Generate UUIDs

To generate a random version 4 UUID for regular identifiers, such as user or database IDs, 
use the [`Uuid.random()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/random.html) function:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart    
    val id = Uuid.random()
    println(id)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

You can also generate a specific version of UUIDs:

> * Use the [`Uuid.generateV4()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v4.html) function to generate a version 4 UUID: 

Reviewer: hmm 🤔 this list is a bit crowded — do you think we could restructure it to work without it?

Author: I removed comparison with the random function, I don't think it's that important tbh.

Reviewer: hmm 🤔 — I think it should be otherwise it might lead to some confusion what's the difference between the two.

  ```kotlin
  import kotlin.uuid.Uuid
  
  fun main() { 
  //sampleStart    
      val id = Uuid.generateV4()
      println(id)
  //sampleEnd    
  }
  ```
  {kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

  This function makes the UUID version explicit, but otherwise works
  the same as the `Uuid.random()` function.

* Use the [`Uuid.generateV7()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7.html) function to generate a version 7 UUID that contains

>   a time-based component and random data:

Reviewer: What do we mean here by random data? 🤔 — I assume the part that makes it unique, but it wouldn't be a UUID without it 🤔 (I might be missing something so please correct me if I'm wrong)

I think this would be clearer if we would say something along the lines:  

Use the [`Uuid.generateV7()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7.html) function to generate a version 7 UUID that includes a timestamp for when it was generated:

Author: Yes, good point, fixed.

  ```kotlin
  import kotlin.uuid.Uuid
  
  fun main() {
  //sampleStart    
      val id = Uuid.generateV7()
      println(id)
  //sampleEnd    
  }
  ```
  {kotlin-runnable="true" kotlin-min-compiler-version="2.3"}

>   Use the `Uuid.generateV7()` function when you need to sort UUIDs by creation time.

Reviewer: Hmm 🤔 I'm a bit worried that this is an overpromise, so WDYT a bit softer version.

The `Uuid.generateV7()` function generates UUIDs that you can sort by their timestamp. or something similar (also avoids repeating Use)

even it gets a bit repetitive I suggest  to differentiate between generate and create words: generate something that the compiler creates - create is something the user creates 🤔

Author: Moved this to the part above the code block

  You can also generate a version 7 UUID for a specific moment in time
  with the [`Uuid.generateV7NonMonotonicAt()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/generate-v7-non-monotonic-at.html) function.

## Parse UUIDs

UUID values are often represented as strings,
such as in URL parameters or database records.

To convert a `String` value to a `Uuid` value,

> use the [`Uuid.parse()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse.htm) function:

Reviewer: ```suggestion
use the [`Uuid.parse()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse.html) function:
```

Author: Good catch!

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart
    val id = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    println(id)
//sampleEnd    
}
```

> {kotlin-runnable="true" kotlin-min-compiler-version="2.0" validate="true"}

Reviewer: As a general note: validate="true" isn't necessary

Author: removed

The `Uuid.parse()` function accepts both the standard hex-and-dash
format and hexadecimal format without dashes.

If the input is invalid, the `Uuid.parse()` function throws an `IllegalArgumentException`:

```kotlin
import kotlin.uuid.Uuid

fun main() { 
//sampleStart    
    val id = Uuid.parse("10")
    println(id)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0" validate="false"}

If your application accepts only one representation, use

> the format specific-functions:

Reviewer: ```suggestion
the format-specific functions:
```

* [`Uuid.parseHex()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex.html) for the hexadecimal string representation without dashes.

For example:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart  
    val standard = Uuid.parseHexDash("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    val compact = Uuid.parseHex("de2bc56cea734f3c8a375a46fdb2d79a")
    
    println(standard)
    println(compact)
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3" validate="true"}

If you have UUIDs from external sources and must handle invalid input safely,
use [`Uuid.parseOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-or-null.html), [`Uuid.parseHexDashOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-dash-or-null.html),
or [`Uuid.parseHexOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/parse-hex-or-null.html). 
These functions return `null` if the input is invalid:

```kotlin

> fun parseId (input: String): Uuid? { 

Reviewer: ```suggestion
fun parseId(input: String): Uuid? { 
```

For example:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart    
    val id = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    
    println(id.toString())
    // de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a
    println(id.toHexDashString())
    // de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a
    println(id.toHexString())
    // de2bc56cea734f3c8a375a46fdb2d79a 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3" validate="true"}

## Compare UUIDs

You can check whether `Uuid` values are equal using the `==` operator.

Kotlin compares values according to the UUID value, not to the textual representation.
For example, two values in different forms are equal if they represent the same 128-bit value:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart    
    val first = Uuid.parse("de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a")
    val second = Uuid.parse("de2bc56cea734f3c8a375a46fdb2d79a")

    println(first == second) 
    // true 
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3" validate="true"}

> This makes the `Uuid` comparison more reliable than string comparing,

Reviewer: ```suggestion
This makes the `Uuid` comparison more reliable than comparing strings,
```

## Work with binary representations

Some APIs, storage formats, and binary protocols do not represent UUIDs as strings. 
Instead, they store the 128-bit UUID value as either:

* A 16-byte array 
* Two 64-bit values

Use these representations when you need to exchange UUIDs with systems that expect binary UUID data.

To convert a UUID to and from 16-byte representation, use the [`.toByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/to-byte-array.html)
and [`Uuid.fromByteArray()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/-uuid/-companion/from-byte-array.html) functions:

```kotlin
import kotlin.uuid.Uuid

fun main() {
//sampleStart 
    val id = Uuid.random()

    val bytes = id.toByteArray()
    val original = Uuid.fromByteArray(bytes)
  
    println(id)
    
    println(bytes)
    println(original)

    println(id == original) 
    // true
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0"}

You can also represent the same 128-bit UUID value as two `Long` values. 
This is useful because Kotlin does not provide a built-in 128-bit integer type. 
The two `Long` values store the UUID in two parts:

> * The `mostSignificantBits` parameter for the first 64 bits of a UUID.

Reviewer: ```suggestion

* The `mostSignificantBits` parameter for the first 64 bits of a UUID.
```

    }
//sampleEnd  
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0" validate="true"}

## Serialize UUIDs

Kotlin supports serialization for `Uuid` values.
Use it to store or transfer a UUID value outside Kotlin code, for example, in JSON APIs or configuration files.

To serialize a `Uuid` value, represent it as a string unless your application requires another format.
The [`kotlinx.serialization`](https://kotlinlang.org/docs/serialization.html) library uses the hex-and-dash format:

```kotlin
//sampleStart 
import kotlin.uuid.Uuid
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class User(
    val id: Uuid,
    val name: String
)

fun main() {
    val user = User(
        id = Uuid.parse("de2bc56cea734f3c8a375a46fdb2d79a"),
        name = "Kotlin"
    )

    println(Json.encodeToString(user))
    // {"id":"de2bc56c-ea73-4f3c-8a37-5a46fdb2d79a","name":"Kotlin"}
}
//sampleEnd
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.3" validate="true"}

## Use UUIDs with Java APIs

To represent UUIDs, Java uses `java.util.UUID` class. On the JVM, Java APIs may

> accept or return this type. Although `java.util.UUID` and `kotlin.unit.UUID` both represent UUIDs,

Reviewer: ```suggestion
accept or return this type. Although `java.util.UUID` and `kotlin.unit.Uuid` both represent UUIDs,
```

they are two distinct types. 

Convert values explicitly to pass UUIDs between Kotlin and Java:

* Convert a Java UUID to Kotlin with the [`.toKotlinUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-kotlin-uuid.html) extension function:

  ```kotlin
  import kotlin.uuid.toKotlinUuid
  
  val kotlinId: Uuid = javaId.toKotlinUuid()
  ```

* Convert a Kotlin UUID to Java with the [`.toJavaUuid()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.uuid/to-java-uuid.html) extension function:

  ```kotlin
  import kotlin.uuid.toJavaUuid
  
  val javaId: java.util.UUID = kotlinId.toJavaUuid()
  ```

These functions allow you to represent your UUID values using `Uuid` at JVM interoperability boundaries.

> > The `java.util.UUID` and `kotlin.unit.UUID` classes are comparable, but the ordering may differ.

Reviewer: ```suggestion
> The `java.util.UUID` and `kotlin.unit.Uuid` classes are comparable, but the ordering may differ.
```
